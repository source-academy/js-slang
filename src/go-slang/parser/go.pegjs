{{
    function extractList(list, index) {
        return list.map(element => element[index]);
    }
    
    function buildList(head, tail, index) {
        return [head].concat(extractList(tail, index));
    }

    function buildInteger(str, base) {
        // note: we discard the "_" delimiters
        return parseInt(str.replaceAll("_", ""), base)
    }

    function buildLiteral(value) {
        return {
            type: "Literal",
            value: value
        };
    }
    
    function buildBinaryExpression(head, tail) {
        return tail.reduce(function(result, element) {
            return {
                type: "BinaryExpression",
                operator: element[1],
                left: result,
                right: element[3]
            };
        }, head);
    }
}}

SourceFile
    = topLevelDecls: TopLevelDeclaration* {
        return { type: "SourceFile", topLevelDecls }
      }

TopLevelDeclaration
    = Declaration

Statement
    = Declaration
    / SimpleStatement

Declaration
    = VariableDeclaration

SimpleStatement
    = ShortVariableDeclaration
    / ExpressionStatement

ExpressionStatement
   = expression: Expression { 
        return { type: "ExpressionStatement", expression: expression } 
     }

Expression
    = RelationalExpression 

PrimaryExpression
    = Identifier
    / Literal
    / "(" _ expression: Expression _ ")" { return expression }

Identifier
    = !Keyword Letter IdentifierPart* { 
        return { type: "Identifier", name: text() } 
      }

Letter
    = UnicodeLetter
    / "_"

IdentifierPart
    = Letter
    / UnicodeDigit

Literal
    = BasicLit
 
BasicLit
    = IntegerLit

IntegerLit "integer"
    = HexInt 
    / OctalInt 
    / BinaryInt 
    / DecimalInt

DecimalInt "decimal"
    = "0" { return buildLiteral(0); } 
 	/ [1-9] DecimalDigit* {
        return buildLiteral(buildInteger(text(), 10))
   	  }

BinaryInt "binary"
    = "0b"i digits:$BinaryDigit+ {
        return buildLiteral(buildInteger(digits, 2))
      }

OctalInt "octal"
    = "0" "o"i? digits:$OctalDigit+ {
        return buildLiteral(buildInteger(digits, 8))
      }

HexInt "hexadecimal"
    = "0x"i digits:$HexDigit+ {
        return buildLiteral(buildInteger(digits, 16))
      }
 
DecimalDigit
    = "_"? [0-9]

BinaryDigit
    = "_"? [0-1]
 
OctalDigit
    = "_"? [0-7]
 
HexDigit
    = "_"? [a-fA-F0-9]
 
UnaryExpression
    = PrimaryExpression
    / operator:UnaryOperator argument:UnaryExpression {
        return {type: "UnaryExpression", operator: operator, argument: argument}
 	  }
 
UnaryOperator
    = "+" 
    / "-"

MultiplicativeExpression
    = head:UnaryExpression
      tail:(_ MultiplicativeOperator _ UnaryExpression)*
      { return buildBinaryExpression(head, tail); }

MultiplicativeOperator
    = "*" 
    / "/" 
    / "%" 

AdditiveExpression
    = head:MultiplicativeExpression
      tail:(_ AdditiveOperator _ MultiplicativeExpression)*
      { return buildBinaryExpression(head, tail); }
 
AdditiveOperator
    = "+" 
    / "-" 
    / "|" 
    / "^"

RelationalExpression
    = head:AdditiveExpression
      tail:(_ RelationalOperator _ AdditiveExpression)*
      { return buildBinaryExpression(head, tail); }

RelationalOperator
    = "=="
    / "!="
    / "<="
    / "<"
    / ">="
    / ">"

VariableDeclaration
    = VAR_TOKEN __ declarations:VarSpec EOS {
        return { type: "VariableDeclaration", ...declarations }
      }

VarSpec
    = ids:IdentifierList _ exprs:("=" _ ExpressionList)? {
        return { declarations: ids, expressions: exprs && exprs[2] }
      }

ShortVariableDeclaration
    = ids:IdentifierList _ ":=" _ exprs:ExpressionList {
        return { type: "VariableDeclaration", declarations: ids, expressions: exprs }
      }

IdentifierList
    = head:Identifier _ tail:(_ "," _ Identifier)* { return buildList(head, tail, 3); }

ExpressionList
    = head:Expression _ tail:(_ "," _ Expression)* { return buildList(head, tail, 3); }

/* Tokens */

BREAK_TOKEN         = "break"          !IdentifierPart
DEFAULT_TOKEN       = "default"        !IdentifierPart
FUNC_TOKEN          = "func"           !IdentifierPart
INTERFACE_TOKEN     = "interface"      !IdentifierPart
SELECT_TOKEN        = "select"         !IdentifierPart
CASE_TOKEN          = "case"           !IdentifierPart
DEFER_TOKEN         = "defer"          !IdentifierPart
GO_TOKEN            = "go"             !IdentifierPart
MAP_TOKEN           = "map"            !IdentifierPart
STRUCT_TOKEN        = "struct"         !IdentifierPart
CHAN_TOKEN          = "chan"           !IdentifierPart
ELSE_TOKEN          = "else"           !IdentifierPart
GOTO_TOKEN          = "goto"           !IdentifierPart
PACKAGE_TOKEN       = "package"        !IdentifierPart
SWITCH_TOKEN        = "switch"         !IdentifierPart
CONST_TOKEN         = "const"          !IdentifierPart
FALLTHROUGH_TOKEN   = "fallthrough"    !IdentifierPart
IF_TOKEN            = "if"             !IdentifierPart
RANGE_TOKEN         = "range"          !IdentifierPart
TYPE_TOKEN          = "type"           !IdentifierPart
CONTINUE_TOKEN      = "continue"       !IdentifierPart
FOR_TOKEN           = "for"            !IdentifierPart
IMPORT_TOKEN        = "import"         !IdentifierPart
RETURN_TOKEN        = "return"         !IdentifierPart
VAR_TOKEN           = "var"            !IdentifierPart

Keyword
    = BREAK_TOKEN    
    / DEFAULT_TOKEN
    / FUNC_TOKEN
    / INTERFACE_TOKEN
    / SELECT_TOKEN
    / CASE_TOKEN
    / DEFER_TOKEN
    / GO_TOKEN
    / MAP_TOKEN
    / STRUCT_TOKEN
    / CHAN_TOKEN
    / ELSE_TOKEN
    / GOTO_TOKEN
    / PACKAGE_TOKEN
    / SWITCH_TOKEN
    / CONST_TOKEN  
    / FALLTHROUGH_TOKEN
    / IF_TOKEN
    / RANGE_TOKEN
    / TYPE_TOKEN    
    / CONTINUE_TOKEN
    / FOR_TOKEN 
    / IMPORT_TOKEN
    / RETURN_TOKEN   
    / VAR_TOKEN   

/* Separators */

EOS
  = _ LineTerminatorSequence?

LineTerminatorSequence "end of line"
  = "\n"
  / "\r\n"
  / "\r"
  / "\u2028"
  / "\u2029"

_  "whitespace" = [ \t\r\n]* // optional whitespace
__ "whitespace" = [ \t\r\n]+ // required whitespace
