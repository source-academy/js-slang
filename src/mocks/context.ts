import * as es from 'estree'

import createContext from '../createContext'
import Closure from '../interpreter/closure'
import { Context, Environment, Variant } from '../types'

export function mockContext(chapter = 1, variant: Variant = 'default'): Context {
  return createContext(chapter, variant)
}

export function mockRuntimeContext(): Context {
  const context = createContext()
  context.runtime = {
    break: false,
    debuggerOn: true,
    isRunning: true,
    environments: [],
    nodes: [
      {
        type: 'Literal',
        loc: {
          start: { line: 1, column: 0 },
          end: { line: 1, column: 1 }
        },
        value: 0,
        raw: '0',
        range: [0, 1]
      }
    ]
  }
  return context
}

export function mockClosure(): Closure {
  return new Closure(
    {
      type: 'FunctionExpression',
      loc: null,
      id: null,
      params: [],
      body: {
        type: 'BlockStatement',
        body: []
      }
    } as es.FunctionExpression,
    {} as Environment,
    {} as Context
  )
}
