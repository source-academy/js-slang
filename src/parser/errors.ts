import { Node, SourceLocation } from 'estree'

import { UNKNOWN_LOCATION } from '../constants'
import { ErrorSeverity, ErrorType, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc ?? UNKNOWN_LOCATION
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
        You are trying to use ${this.nodeType}, which is not allowed (yet).
      `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      case 'ImportNamespaceSpecifier':
        return 'Namespace imports'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}
