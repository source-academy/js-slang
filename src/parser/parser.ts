/* tslint:disable:max-classes-per-file */
import {
  Options as AcornOptions,
  parse as acornParse,
  parseExpressionAt as acornParseAt,
  Position
} from 'acorn'
import { parse as acornLooseParse } from 'acorn-loose'
import { ancestor, AncestorWalkerFn } from 'acorn-walk/dist/walk'
import * as es from 'estree'
import { Context, ErrorSeverity, ErrorType, Rule, SourceError } from '../types'
import { stripIndent } from '../utils/formatters'
import rules from './rules'
import syntaxBlacklist from './syntaxBlacklist'

export class DisallowedConstructError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.formatNodeType(this.node.type)
  }

  get location() {
    return this.node.loc!
  }

  public explain() {
    return `${this.nodeType} are not allowed`
  }

  public elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not allowed (yet).
    `
  }

  /**
   * Converts estree node.type into english
   * e.g. ThisExpression -> 'this' expressions
   *      Property -> Properties
   *      EmptyStatement -> Empty Statements
   */
  private formatNodeType(nodeType: string) {
    switch (nodeType) {
      case 'ThisExpression':
        return "'this' expressions"
      case 'Property':
        return 'Properties'
      default: {
        const words = nodeType.split(/(?=[A-Z])/)
        return words.map((word, i) => (i === 0 ? word : word.toLowerCase())).join(' ') + 's'
      }
    }
  }
}

export class FatalSyntaxError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation, public message: string) {}

  public explain() {
    return this.message
  }

  public elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements SourceError {
  public type = ErrorType.SYNTAX
  public severity = ErrorSeverity.ERROR
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Missing semicolon at the end of statement'
  }

  public elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements SourceError {
  public type: ErrorType.SYNTAX
  public severity: ErrorSeverity.WARNING
  public constructor(public location: es.SourceLocation) {}

  public explain() {
    return 'Trailing comma'
  }

  public elaborate() {
    return 'Please remove the trailing comma'
  }
}

export function parseAt(source: string, num: number) {
  let theNode: acorn.Node | undefined
  try {
    theNode = acornParseAt(source, num)
  } catch (error) {
    return undefined
  }
  return theNode
}

export function parse(source: string, context: Context, fallbackToLooseParse: boolean = false) {
  let program: es.Program | undefined
  try {
    program = (acornParse(source, createAcornParserOptions(context)) as unknown) as es.Program
    ancestor(program as es.Node, walkers, undefined, context)
  } catch (error) {
    if (error instanceof SyntaxError) {
      // tslint:disable-next-line:no-any
      const loc = (error as any).loc
      const location = {
        start: { line: loc.line, column: loc.column },
        end: { line: loc.line, column: loc.column + 1 }
      }
      context.errors.push(new FatalSyntaxError(location, error.toString()))
    } else {
      throw error
    }
  }
  const hasErrors = context.errors.find(m => m.severity === ErrorSeverity.ERROR)
  if (program && !hasErrors) {
    return program
  } else if (fallbackToLooseParse) {
    return looseParse(source, context)
  } else {
    return undefined
  }
}

const createAcornParserOptions = (context: Context): AcornOptions => ({
  sourceType: 'module',
  ecmaVersion: 6,
  locations: true,
  // tslint:disable-next-line:no-any
  onInsertedSemicolon(end: any, loc: any) {
    context.errors.push(
      new MissingSemicolonError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  // tslint:disable-next-line:no-any
  onTrailingComma(end: any, loc: Position) {
    context.errors.push(
      new TrailingCommaError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  }
})

// Names-extractor needs comments
export function parseForNames(source: string): [es.Program, acorn.Comment[]] {
  let comments: acorn.Comment[] = []
  const options: AcornOptions = {
    sourceType: 'module',
    ecmaVersion: 6,
    locations: true,
    onComment: comments
  }
  let program: es.Program | undefined
  try {
    program = (acornParse(source, options) as unknown) as es.Program
  } catch {
    comments = []
    program = acornLooseParse(source, options)
  }

  return [program, comments]
}

export function looseParse(source: string, context: Context) {
  const program = (acornLooseParse(
    source,
    createAcornParserOptions(context)
  ) as unknown) as es.Program
  return program
}

function createWalkers(
  allowedSyntaxes: { [nodeName: string]: number },
  parserRules: Rule<es.Node>[]
) {
  const newWalkers = new Map<string, AncestorWalkerFn<Context>>()
  const visitedNodes = new Set<es.Node>()

  // Provide callbacks checking for disallowed syntaxes, such as case, switch...
  const syntaxPairs = Object.entries(allowedSyntaxes)
  syntaxPairs.map(pair => {
    const syntax = pair[0]
    newWalkers.set(syntax, (node: es.Node, context: Context, ancestors: [es.Node]) => {
      if (!visitedNodes.has(node)) {
        visitedNodes.add(node)

        if (context.chapter < allowedSyntaxes[node.type]) {
          context.errors.push(new DisallowedConstructError(node))
        }
      }
    })
  })

  // Provide callbacks checking for rule violations, e.g. no block arrow funcs, non-empty lists...
  parserRules.forEach(rule => {
    const checkers = rule.checkers
    const syntaxCheckerPair = Object.entries(checkers)
    syntaxCheckerPair.forEach(pair => {
      const syntax = pair[0]
      const checker = pair[1]
      const oldCheck = newWalkers.get(syntax)!
      const newCheck = (node: es.Node, context: Context, ancestors: es.Node[]) => {
        if (typeof rule.disableOn !== 'undefined' && context.chapter >= rule.disableOn) {
          return
        }
        const errors = checker(node, ancestors)
        errors.forEach(e => context.errors.push(e))
      }
      newWalkers.set(syntax, (node: es.Node, context: Context<any>, ancestors) => {
        oldCheck(node, context, ancestors)
        newCheck(node, context, ancestors)
      })
    })
  })

  return mapToObj(newWalkers)
}

const mapToObj = (map: Map<string, any>) =>
  Array.from(map).reduce((obj, [k, v]) => Object.assign(obj, { [k]: v }), {})

const walkers: { [name: string]: AncestorWalkerFn<Context> } = createWalkers(syntaxBlacklist, rules)
