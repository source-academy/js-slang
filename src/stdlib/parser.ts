import { oneLine } from 'common-tags'
import * as es from 'estree'

import { parse as sourceParse } from '../parser'
import { Context, Value } from '../types'
import { vector_to_list } from './list'

class ParseError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ParseError'
  }
}

function unreachable() {
  // tslint:disable-next-line:no-console
  console.error(oneLine`
    UNREACHABLE CODE REACHED!
    Please file an issue at
    https://github.com/source-academy/js-slang/issues
    if you see this.
  `)
}

type ASTTransformers = Map<string, (node: es.Node) => Value>

let transformers: ASTTransformers
transformers = new Map([
  [
    'Program',
    (node: es.Node) => {
      node = node as es.Program
      return vector_to_list(node.body.map(transform))
    }
  ],

  [
    'BlockStatement',
    (node: es.Node) => {
      node = node as es.BlockStatement
      return {
        tag: 'block',
        body: vector_to_list(node.body.map(transform))
      }
    }
  ],

  [
    'ExpressionStatement',
    (node: es.Node) => {
      node = node as es.ExpressionStatement
      return transform(node.expression)
    }
  ],

  [
    'IfStatement',
    (node: es.Node) => {
      node = node as es.IfStatement
      return {
        tag: 'conditional_statement',
        predicate: transform(node.test),
        consequent: transform(node.consequent),
        alternative: transform(node.alternate as es.Statement)
      }
    }
  ],

  [
    'FunctionDeclaration',
    (node: es.Node) => {
      node = node as es.FunctionDeclaration
      return {
        tag: 'constant_declaration',
        name: transform(node.id as es.Identifier),
        value: {
          tag: 'function_definition',
          parameters: vector_to_list(node.params.map(transform)),
          body: vector_to_list(node.body.body.map(transform))
        }
      }
    }
  ],

  [
    'VariableDeclaration',
    (node: es.Node) => {
      node = node as es.VariableDeclaration
      if (node.kind === 'let') {
        return {
          tag: 'variable_declaration',
          name: transform(node.declarations[0].id),
          value: transform(node.declarations[0].init as es.Expression)
        }
      } else if (node.kind === 'const') {
        return {
          tag: 'constant_declaration',
          name: transform(node.declarations[0].id),
          value: transform(node.declarations[0].init as es.Expression)
        }
      } else {
        unreachable()
        throw new ParseError('Invalid declaration kind')
      }
    }
  ],

  [
    'ReturnStatement',
    (node: es.Node) => {
      node = node as es.ReturnStatement
      return {
        tag: 'return_statement',
        expression: transform(node.argument as es.Expression)
      }
    }
  ],

  [
    'CallExpression',
    (node: es.Node) => {
      node = node as es.CallExpression
      return {
        tag: 'application',
        operator: transform(node.callee),
        operands: vector_to_list(node.arguments.map(transform))
      }
    }
  ],

  [
    'UnaryExpression',
    (node: es.Node) => {
      node = node as es.UnaryExpression
      const loc = node.loc as es.SourceLocation
      return {
        tag: 'application',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: loc.start,
            end: { line: loc.start.line, column: loc.start.column + 1 }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.argument)])
      }
    }
  ],

  [
    'BinaryExpression',
    (node: es.Node) => {
      node = node as es.BinaryExpression
      const loc = node.right.loc as es.SourceLocation
      return {
        tag: 'application',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: { line: loc.start.line, column: loc.start.column - 1 },
            end: { line: loc.start.line, column: loc.start.column }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.left), transform(node.right)])
      }
    }
  ],

  [
    'LogicalExpression',
    (node: es.Node) => {
      node = node as es.LogicalExpression
      const loc = node.right.loc as es.SourceLocation
      return {
        tag: 'boolean_operation',
        operator: {
          tag: 'name',
          name: node.operator,
          loc: {
            start: { line: loc.start.line, column: loc.start.column - 1 },
            end: { line: loc.start.line, column: loc.start.column }
          } as es.SourceLocation
        },
        operands: vector_to_list([transform(node.left), transform(node.right)])
      }
    }
  ],

  [
    'ConditionalExpression',
    (node: es.Node) => {
      node = node as es.ConditionalExpression
      return {
        tag: 'conditional_expression',
        predicate: transform(node.test),
        consequent: transform(node.consequent),
        alternative: transform(node.alternate)
      }
    }
  ],

  [
    'ArrowFunctionExpression',
    (node: es.Node) => {
      node = node as es.ArrowFunctionExpression
      return {
        tag: 'function_definition',
        parameters: vector_to_list(node.params.map(transform)),
        body: {
          tag: 'return_statement',
          expression: transform(node.body as es.Expression),
          loc: node.body.loc
        }
      }
    }
  ],

  [
    'Identifier',
    (node: es.Node) => {
      node = node as es.Identifier
      return {
        tag: 'name',
        name: node.name
      }
    }
  ],

  [
    'Literal',
    (node: es.Node) => {
      node = node as es.Literal
      return node.value
    }
  ],

  [
    'ArrayExpression',
    (node: es.Node) => {
      node = node as es.ArrayExpression
      if (node.elements.length === 0) {
        return {
          tag: 'empty_list'
        }
      } else {
        return {
          tag: 'array_expression',
          elements: vector_to_list(node.elements.map(transform))
        }
      }
    }
  ],

  [
    'AssignmentExpression',
    (node: es.Node) => {
      node = node as es.AssignmentExpression
      if (node.left.type === 'Identifier') {
        return {
          tag: 'assignment',
          name: transform(node.left as es.Identifier),
          value: transform(node.right)
        }
      } else if (node.left.type === 'MemberExpression') {
        return {
          tag: 'property_assignment',
          object: transform(node.left as es.Expression),
          value: transform(node.right)
        }
      } else {
        unreachable()
        throw new ParseError('Invalid assignment')
      }
    }
  ],

  [
    'ForStatement',
    (node: es.Node) => {
      node = node as es.ForStatement
      return {
        tag: 'for_loop',
        initialiser: transform(node.init as es.VariableDeclaration | es.Expression),
        predicate: transform(node.test as es.Expression),
        finaliser: transform(node.update as es.Expression),
        statements: transform(node.body)
      }
    }
  ],

  [
    'WhileStatement',
    (node: es.Node) => {
      node = node as es.WhileStatement
      return {
        tag: 'while_loop',
        predicate: transform(node.test),
        statements: transform(node.body)
      }
    }
  ],

  [
    'BreakStatement',
    (node: es.Node) => {
      node = node as es.BreakStatement
      return {
        tag: 'break_statement'
      }
    }
  ],

  [
    'ContinueStatement',
    (node: es.Node) => {
      node = node as es.ContinueStatement
      return {
        tag: 'continue_statement'
      }
    }
  ],

  [
    'ObjectExpression',
    (node: es.Node) => {
      node = node as es.ObjectExpression
      return {
        tag: 'object_expression',
        pairs: vector_to_list(node.properties.map(transform))
      }
    }
  ],

  [
    'MemberExpression',
    (node: es.Node) => {
      node = node as es.MemberExpression
      if (node.computed) {
        return {
          tag: 'property_access',
          object: transform(node.object),
          property: transform(node.property)
        }
      } else {
        const prop = node.property as es.Identifier
        return {
          tag: 'property_access',
          object: transform(node.object),
          property: prop.name
        }
      }
    }
  ],

  [
    'Property',
    (node: es.Node) => {
      node = node as es.Property
      if (node.key.type === 'Literal') {
        return [node.key.value, transform(node.value)]
      } else if (node.key.type === 'Identifier') {
        return [node.key.name, transform(node.value)]
      } else {
        unreachable()
        throw new ParseError('Invalid property key type')
      }
    }
  ]
])

function transform(node: es.Node) {
  if (transformers.has(node.type)) {
    const transformer = transformers.get(node.type) as (n: es.Node) => Value
    const transformed = transformer(node)
    // Attach location information
    if (
      transformed !== null &&
      transformed !== undefined &&
      typeof transformed === 'object' &&
      transformed.tag !== undefined
    ) {
      transformed.loc = node.loc
    }
    return transformed
  } else {
    unreachable()
    throw new ParseError('Cannot transform unknown type: ' + node.type)
  }
}

export function parse(x: string, context: Context): Value {
  let program
  program = sourceParse(x, context)
  if (context.errors.length > 0) {
    throw new ParseError(context.errors[0].explain())
  }

  if (program !== undefined) {
    return transform(program)
  } else {
    unreachable()
    throw new ParseError('Invalid parse')
  }
}
