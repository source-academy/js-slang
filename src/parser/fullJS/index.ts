import { parse } from 'acorn'
import type { Program } from 'estree'

import type { Context } from '../..'
import { FatalSyntaxError } from '../errors'
import type { AcornOptions, Parser } from '../types'
import { positionToSourceLocation } from '../utils'

export class FullJSParser implements Parser<AcornOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<AcornOptions>,
    throwOnError?: boolean
  ): Program | null {
    try {
      return parse(programStr, {
        sourceType: 'module',
        ecmaVersion: 'latest',
        locations: true,
        ...options
      }) as unknown as Program
    } catch (error) {
      if (error instanceof SyntaxError) {
        error = new FatalSyntaxError(positionToSourceLocation((error as any).loc), error.toString())
      }

      if (throwOnError) throw error
      context.errors.push(error)
    }

    return null
  }

  validate(_ast: Program, _context: Context, _throwOnError: boolean): boolean {
    return true
  }

  toString(): string {
    return 'FullJSParser'
  }
}
