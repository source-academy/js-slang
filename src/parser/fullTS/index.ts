import { parse as babelParse } from '@babel/parser'
import { createProjectSync, ts } from '@ts-morph/bootstrap'
import { Program } from 'estree'

import { Context } from '../..'
import * as TypedES from '../../typeChecker/tsESTree'
import { removeTSNodes } from '../../typeChecker/typeErrorChecker'
import { FatalSyntaxError } from '../errors'
import { transformBabelASTToESTreeCompliantAST } from '../source/typed/utils'
import { AcornOptions, Parser } from '../types'
import { defaultBabelOptions, positionToSourceLocation } from '../utils'

const IMPORT_TOP_LEVEL_ERROR =
  'An import declaration can only be used at the top level of a namespace or module.'
const START_OF_MODULE_ERROR = 'Cannot find module '

export class FullTSParser implements Parser<AcornOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<AcornOptions>,
    throwOnError?: boolean
  ): Program | null {
    let code = ''
    // Add builtins to code
    // Each declaration is replaced with a single constant declaration with type `any`
    // to reduce evaluation time
    for (const builtin of context.nativeStorage.builtins) {
      code += `const ${builtin[0]}: any = 1\n`
    }
    // Add prelude functions to code
    // Each declaration is replaced with a single constant declaration with type `any`
    // to reduce evaluation time
    if (context.prelude) {
      const preludeFns = context.prelude.split('\nfunction ').slice(1)
      preludeFns.forEach(fnString => {
        const fnName = fnString.split('(')[0]
        // Functions in prelude that start with $ are not added
        if (fnName.startsWith('$')) {
          return
        }
        code += `const ${fnName}: any = 1\n`
      })
    }
    // Get line offset
    const lineOffset = code.split('\n').length - 1

    // Add program string to code string,
    // wrapping it in a block to allow redeclaration of variables
    code = code + '{' + programStr + '}'
    // Initialize file to analyze
    const project = createProjectSync({ useInMemoryFileSystem: true })
    const filename = 'program.ts'
    project.createSourceFile(filename, code)

    // Get TS diagnostics from file, formatted as TS error string
    const diagnostics = ts.getPreEmitDiagnostics(project.createProgram())
    const formattedString = project.formatDiagnosticsWithColorAndContext(diagnostics)

    // Reformat TS error string to Source error by getting line number using regex
    // This is because logic to retrieve line number is only present in
    // formatDiagnosticsWithColorAndContext and cannot be called directly
    const lineNumRegex = /(?<=\[7m)\d+/
    diagnostics.forEach(diagnostic => {
      const message = diagnostic.messageText.toString()
      // Ignore errors regarding imports
      // as TS does not have information about Source modules
      if (message === IMPORT_TOP_LEVEL_ERROR || message.startsWith(START_OF_MODULE_ERROR)) {
        return
      }
      const lineNumRegExpArr = lineNumRegex.exec(formattedString.split(message)[1])
      const lineNum = (lineNumRegExpArr === null ? 0 : parseInt(lineNumRegExpArr[0])) - lineOffset
      // Ignore any errors that occur in builtins/prelude (line number <= 0)
      if (lineNum <= 0) {
        return
      }
      const position = { line: lineNum, column: 0, offset: 0 }
      context.errors.push(new FatalSyntaxError(positionToSourceLocation(position), message))
    })

    if (context.errors.length > 0) {
      return null
    }

    // Parse code into Babel AST, which supports type syntax
    const ast = babelParse(programStr, {
      ...defaultBabelOptions,
      sourceFilename: options?.sourceFile,
      errorRecovery: throwOnError ?? true
    })

    if (ast.errors.length) {
      ast.errors
        .filter(error => error instanceof SyntaxError)
        .forEach(error => {
          context.errors.push(
            new FatalSyntaxError(
              positionToSourceLocation((error as any).loc, options?.sourceFile),
              error.toString()
            )
          )
        })

      return null
    }

    // Transform Babel AST into ESTree AST
    const typedProgram: TypedES.Program = ast.program as TypedES.Program
    const transpiledProgram: Program = removeTSNodes(typedProgram)
    transformBabelASTToESTreeCompliantAST(transpiledProgram)

    return transpiledProgram
  }

  validate(_ast: Program, _context: Context, _throwOnError: boolean): boolean {
    return true
  }

  toString(): string {
    return 'FullTSParser'
  }
}
