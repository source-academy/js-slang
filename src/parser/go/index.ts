import { FatalSyntaxError } from '../errors'
import { Parser } from '../types'
import { parse } from '../../go-slang/parser/go.js'

export class GoParser implements Parser<any> {
  parse(programStr: string, context: any, options?: any, throwOnError?: boolean): any {
    try {
      return parse(programStr)
    } catch (error) {
      const location = error.location
      error = new FatalSyntaxError(
        {
          start: { line: location.start.line, column: location.start.column },
          end: { line: location.end.line, column: location.end.column },
          source: location.source
        },
        error.toString()
      )

      if (throwOnError) throw error
      context.errors.push(error)
    }

    return null
  }
  validate(ast: any, context: any, throwOnError?: boolean): boolean {
    return true
  }
}
