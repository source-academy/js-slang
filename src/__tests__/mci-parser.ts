import { mockContext } from '../mocks/context'
//@ts-ignore
import { Parser } from '../stdlib/parser.js'

const parser = new Parser();

test('Parse an arrow function', () => {
  const context = mockContext()
  const program = parser.parse('x => x + 1;', context)
  expect(program).toMatchSnapshot()
})

test('Parse arrow function assignment', () => {
  const context = mockContext()
  const program = parser.parse('const y = x => x + 1;', context)
  expect(program).toMatchSnapshot()
})

