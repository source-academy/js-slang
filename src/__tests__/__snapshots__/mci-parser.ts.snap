// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse an arrow function 1`] = `
Array [
  Object {
    "body": Object {
      "expression": Object {
        "line": 0,
        "operands": Array [
          Object {
            "line": 0,
            "name": "x",
            "tag": "name",
          },
          Array [
            1,
            Array [],
          ],
        ],
        "operator": Object {
          "line": 0,
          "name": "+",
          "tag": "name",
        },
        "tag": "application",
      },
      "line": 0,
      "tag": "return_statement",
    },
    "line": 0,
    "location": Object {
      "end_col": 0,
      "end_line": 1,
      "start_col": 0,
      "start_line": 1,
    },
    "parameters": Array [
      Object {
        "line": 0,
        "name": "x",
        "tag": "name",
      },
      Array [],
    ],
    "tag": "function_definition",
  },
  Array [],
]
`;

exports[`Parse arrow function assignment 1`] = `
Array [
  Object {
    "line": 0,
    "name": "y",
    "tag": "constant_declaration",
    "value": Object {
      "body": Object {
        "expression": Object {
          "line": 0,
          "operands": Array [
            Object {
              "line": 0,
              "name": "x",
              "tag": "name",
            },
            Array [
              1,
              Array [],
            ],
          ],
          "operator": Object {
            "line": 0,
            "name": "+",
            "tag": "name",
          },
          "tag": "application",
        },
        "line": 0,
        "tag": "return_statement",
      },
      "line": 0,
      "location": Object {
        "end_col": 10,
        "end_line": 1,
        "start_col": 10,
        "start_line": 1,
      },
      "parameters": Array [
        Object {
          "line": 0,
          "name": "x",
          "tag": "name",
        },
        Array [],
      ],
      "tag": "function_definition",
    },
  },
  Array [],
]
`;
